src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.PlatformVersion.cs

```diff
@@ -0,0 +1,131 @@
+
+        /// <summary>
+        /// Check for the OS with a &lt; version comparison. Used to guard APIs that were obsoleted or removed in the given OS release.
+        /// </summary>
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major)
+            => IsOSPlatform(osPlatform) && Environment.OSVersion.Version.Major < major;
+
+        /// <summary>
+        /// Check for the OS with a &lt; version comparison. Used to guard APIs that were obsoleted or removed in the given OS release.
+        /// </summary>
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major, int minor)
+            => IsOSPlatform(osPlatform) && !IsOSVersionOrLater(major, minor, int.MinValue, int.MinValue);
+
+        /// <summary>
+        /// Check for the OS with a &lt; version comparison. Used to guard APIs that were obsoleted or removed in the given OS release.
+        /// </summary>
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major, int minor, int build)
+            => IsOSPlatform(osPlatform) && !IsOSVersionOrLater(major, minor, build, int.MinValue);
+
+        /// <summary>
+        /// Check for the OS with a &lt; version comparison. Used to guard APIs that were obsoleted or removed in the given OS release.
+        /// </summary>
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major, int minor, int build, int revision)
+            => IsOSPlatform(osPlatform) && !IsOSVersionOrLater(major, minor, build, revision);
+
+        private static bool IsOSVersionOrLater(int major, int minor, int build, int revision)
+        {
+            Version current = Environment.OSVersion.Version;
+            if (current.Major != major)
+            {
+                return current.Major > major;
+            }
+            if (current.Minor != minor)
+            {
+                return current.Minor > minor;
+            }
+            if (current.Build != build)
+            {
+                return current.Build > build;
+            }
+
+            return current.Revision >= revision;
+        }
+
+        private static (OSPlatform, Version) Parse(string platformName)
+        {
+            if (platformName == null)
+            {
+                throw new ArgumentNullException(nameof(platformName));
+            }
+            if (platformName.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_EmptyValue, nameof(platformName));
+            }
+
+            // iterate from the begining, as digits in the middle of the names are not supported by design
+            for (int i = 0; i < platformName.Length; i++)
+            {
+                if (char.IsDigit(platformName[i]))
+                {
+                    if (i > 0 && Version.TryParse(platformName.AsSpan(i), out Version? version))
+                    {
+                        return (OSPlatform.Create(platformName.Substring(0, i)), version);
```

Comment:
@adamsitnik - with https://github.com/dotnet/runtime/pull/39209/files#diff-18334e5619174ba5d33e2f1dafcf5fb2R45, we added a remark

> If you plan to call this method frequently, please consider caching its result.

Did we just create a perf trap with `IsOSPlatformOrLater(string platformName)` and `IsOSPlatformEarlierThan(string platformName)`?

Reply:
I think we ended up with the best possible compromise between performance and usability. But the conversation we had around it also included ensuring we have `Debug.Assert` support. That will allow especially perf-conscious developers to use a `Debug.Assert(RuntimeInformation.IsOSPlatformOrLater(platformName));` call before their platform-specific code instead of needing the `if` statement.  Then the condition will only be checked in debug builds and can be trimmed away in release builds.